### ISSUE TITLE
Support Caching as a built-in Feature

### ISSUE THREAD

**Issue by ardalis** [dotnet/efcore#27735]
Most web applications are "read mostly" meaning they query data much more frequently than they change it. This means they benefit immensely from caching. Back in 2005, ASP.NET introduced the SqlCacheDependency, which supported to modes of operation:
1) Triggers + Polling informed the application when changes occurred to a table that was being cached
2) App subscribes to SQL Server and is notified when changes would impact a given query. Provides row/column level cache invalidation. This solution is far more efficient and works well with web farms.

This article briefly describes the feature:
https://docs.microsoft.com/en-us/previous-versions/aspnet/ms178604(v=vs.100)

These features provided dramatic performance improvements with minimal effort 17 years ago but have never been carried forward or made compatible with EF (nor EF Core) as far as I'm aware. I've requested this feature periodically since EF was created but so far without success.

I would like to be able to enable caching on a per-entity basis as part of EF configuration. This caching should automatically be updated when the underlying data changes, using one of the above techniques or some novel approach. I don't need every caching feature under the sun (durations, sliding windows, etc etc) - I just want to be able to reduce the number of needlessly repetitive queries on lookup tables and other read-mostly data.

**Comment by roji** [comment]
@ardalis I'll look into this in more detail, however:

> I'm unaware of any hooks that allow changes to whether, when, an how EF chooses to issue queries to its data store. This is data-level caching, below the level of entities and change-tracking. If there are hooks that someone can use to add a SqlCacheDependency to a SqlCommand used by EF please show me and I'll gladly create a POC for doing this myself. I just don't think that low level implementation has ever been exposed for third party extensions.

EF Core does expose hooks for [command interceptors](https://docs.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors#database-interception), which allow you to intercept command execution and e.g. provide cached results; this could be the basis of writing an in-memory caching layer for EF Core, and also to integrate it with SqlCacheDependencies. Let me know if this is helpful.

**Comment by roji** [comment]
Note that unless I'm mistaken, SqlCacheDependency is part of System.Web, and only available on .NET Framework (not on .NET Core/.NET 5+). Of course, it's still possible to set up triggers/polling to invalidate your caching layer, but it wouldn't be possible to use newer versions of EF Core with SqlCacheDependency specifically.

**Comment by ardalis** [comment]
Right, I'd anticipate the EF team creating their own modernized version of the SqlCacheDependency as one of the key parts of implementing this feature.

**Comment by ajcvickers** [comment]
@ardalis The interesting thing about this (and other second-level cache-like things) is that they were super-hot 15 years ago, but I've been waiting for someone to ask for it on EF Core and, as far as I am aware, this is the first time it has been asked for. I don't really have an explanation for why this is, but it certainly isn't something people are clamoring for anymore.

**Comment by ardalis** [comment]
Maybe I was the only one who thought it was a great feature back in the day!

I think it may be a case of the whole "If I'd delivered what was asked for, I'd have made faster horses." People want their data access to be faster and more efficient. They may not know that this feature DOES EXACTLY THAT. So, while maybe not many people have talked about this specific feature (I know I have but it's been a few years), I suspect you do get regular requests to make EF faster and more efficient.

If you were to take a sample application that is read mostly, like any content-based website, and run some load tests against it (while showing load on the database), and then flip this feature on and show the difference, I think you'd see a lot of interested customers. And it would make a great BUILD demo.

**Comment by ErikEJ** [comment]
@ardalis EF Core has great extensibility

**Comment by roji** [comment]
@ardalis for features where there's no real value in including in core and it's just as good to implement as an extension, we general tend the latter - on principle we try to keep extensibility good enough so that stuff can be done outside of the core, by the community etc. (and indeed, there's a really nice ecosystem of community-maintained EF Core extensions). I think this is especially valid in this case, since a full solution would integrate with a specific in-app caching layer, which could be different across applications.

But let us know how it goes!

**Comment by michalsznajder** [comment]
FYI There is [EFCoreSecondLevelCacheInterceptor](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor) that uses interceptors to implement second level caching.

However it also uses some voodoo magic to parse raw SQL queries that seems to be quite fragile. And is not async friendly.

**Comment by michalsznajder** [comment]
There is also abandoned [EtityFrameworkCore.Cacheable](https://github.com/SteffenMangold/EntityFrameworkCore.Cacheable) that works by overriding query compiler. Looks like there is  no invalidation on CRUD operations.

There was also discussion at https://github.com/nopSolutions/nopCommerce/issues/239

**Comment by programcsharp** [comment]
@ardalis Did you ever get any progress on caching via intercepters?

@ajcvickers / @roji The reason you haven't seen second-level caching asked for may be because like @ardalis, many people want it but some folks don't know to ask for it and others (like myself) who know what second level caching is and the importance of it when you're scaling up (vs out) won't use EF until there's a built-in second-level caching story.

This is something that NHibernate had decades ago -- https://nhibernate.info/doc/nhibernate-reference/caches.html -- so IMO it's table stakes for an ORM. You need to be able to set a bit and automatically cache everything either in memory (for a single server) or in redis (for a cluster), and have invalidation managed automatically (ideally with the capability to invalidate queries for a user at once, not entire entities).

Also IMO, SqlCacheDependency is a different thing -- there could be hooks to invalidate based on it as well, but a good second level caching solution should be able to automatically cache all queries and entities and invalidate on any changes made (because it knows what changed) internally, without a SQL server specific feature.

@ErikEJ Second level caching is not something that can really be added as an extension because of how low level the hooks needed to do it are, and all of the possible interactions between invalidation, transactions, etc. You can't just cache entities directly, to do it right you really need to cache the field values and be able to reconstitute them into objects so that you can maintain guarantees about single instance per context etc. To be able to trust second level caching, it needs to be built-in and supported by the framework.

@michalsznajder Yes, there are a couple of people who have tried to do second level cache extensions, but that just demonstrates how hard this is to add without framework support -- they're all either brittle and buggy or abandoned.

Second level caching storage providers could certainly be extensions -- in memory, redis, any number of other caching services -- but the underlying second level caching system itself needs to be built in.

**Comment by ardalis** [comment]
@programcsharp No, didn't get it working. Have been using solutions like CachedRepository decorator rather than relying on EF directly.

**Comment by roji** [comment]
> the underlying second level caching system itself needs to be built in.

This is something I'm still not quite convinced by. One of your arguments above for integration into EF is that the cache needs to hold raw values rather than entity instances; I definitely agree with that (if only because the cache may need to be external, e.g. redis), but I'm not sure why that can't be done via an interceptor... Our command interception [sample](https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors#example-advanced-command-interception-for-caching) shows exactly this: instead of executing the query, simply return a reader that would wrap results coming out of the cache. Is there something fundamentally problematic with this interception approach that requires integration into EF?

Stepping back, the data-layer caching discussed above does have some disadvantages. For one thing, in many situations it's beneficial to cache above the data access layer, e.g. web pages, web page fragments, or data computed/transformed based on the data returned from the database. That doesn't mean that caching database results directly doesn't make sense in many cases, just that a blanket "turn on 2nd-level caching for all queries" isn't necessarily a good thing.

Similarly, there's the question of *what* to cache; just activating caching on all queries is likely not a good idea in the general case, since lots of memory may be wasted on very cold queries. Querying code can indeed indicate which queries should be cached (the interceptor example above does this via a tag), but sometimes you may even want to cache based on the specific results.

All this isn't to say that a generalized caching layer for your database queries doesn't make sense; but I do feel that caching is too complex an issue to just "turn on" and forget about; it requires careful consideration of which layer or (layers) should implement the caching, exactly what to cache and what not, and of course, how invalidation works.

Finally, at the end of the day, as @ajcvickers mentioned above, we haven't seen much user interest in this. You've written above that maybe users want better speed but don't know that they want caching. That may very well be true, but this kind of argument can be used on almost anything; I also have the intuition that caching is simple and well-known enough that many/most users can understand it and ask for that specifically. "More speed" specifically is a bit misleading; even if we added caching and what-not, I suspect users would continue asking for "more speed" (it's just something that's always wanted).

**Comment by alienwareone** [comment]
> FYI There is [EFCoreSecondLevelCacheInterceptor](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor) that uses interceptors to implement second level caching.
> 
> However it also uses some voodoo magic to parse raw SQL queries that seems to be quite fragile.

I agree. Though I've been using [EFCoreSecondLevelCacheInterceptor](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor) in production for years now and never had any issues.

What it's lacking is support for specific types (e.g. [Spatial](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/4), [HierarchyId](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/170)).
This would require to make internal constructors public for SqlDataReader ([see issue](https://github.com/dotnet/SqlClient/issues/465)).

**Comment by programcsharp** [comment]
@roji Different caching patterns are beneficial for different usage scenarios. Second level caching is needed for that silent boring middle market that's missed in a lot of use case documents -- the line of business app with 10k users, all accessing their own individual data, all hosted on one scaled up database server with a small cluster of webservers. The traditional Microsoft bread and butter of the small and medium sized business.

This is a use case that's much missed by EF (and initially was by a lot of .NET core generally, but that's gotten better over the past few years).

When you're small, writing an MVP with 30 tables, don't have much traffic, or are writing a tutorial, pretty much anything works. When you're doing a flashy synthetic benchmark, a social network app where most of the users are reading the same subset of data and doing minimal writing, or a scale out sharded design, then yes, caching above the data access layer is what you want. You have a couple of hot data sets that are accessed by most users, and a few others that are accessed by smaller subsets but still a large proportion, and you can get a great boost from read only caching of fragments, precomputed data, or even entire pages and you don't much care if the data is a few seconds stale.

When you've got a line of business/SaaS app with lot of users each accessing their own data, 500 or 1000 different tables, and there's a mix of reads and writes (still mostly reads, but writes to their own data as well), that's when you need a standard second level cache. It needs to be something that Just Works, from all of your webservers and service hosts, hitting a cache server like redis, caching reads and automatically invalidating affected items on write. 

You don't have a small hotspot to optimize, you want to optimize general access and offload frequent reads for each of your 10000 users from the database. You opt out of caching large resultsets and other inefficiencies, obviously, but least recently used eviction and expiration makes "cache all of the things by default, opt out of some, extra cache for others" very efficient.

Anyways, that's my $0.2 (inflation 😁) on why second level caching is important and why we (and probably a lot of the less vocal boring companies that make up a ton of the ongoing software that's running the world) can't use EF until it's built-in and supported. I respect your viewpoint as well -- focus is important -- and hopefully this helps clarify where I'm coming from.

**Comment by roji** [comment]
@programcsharp I'm still missing why this has to be built into EF Core; we have at least some evidence that people are using extensions such as EFCoreSecondLevelCacheInterceptor with success. We have to be careful not to fall into the trap of "it all has to come from Microsoft".

**Comment by programcsharp** [comment]
I'm glad you see the use case. Second level caching is imperative for some users.

The "has to" is relative, of course. Nobody is forcing Microsoft to include this feature, but there are some subset of users such as myself and many others like me, who just will not use EF if it doesn't have this key feature built-in and supported.

I've been watching the EF second level caching libraries for years and see them getting started and abandoned (https://github.com/SteffenMangold/EntityFrameworkCore.Cacheable, https://github.com/moozzyk/EFCache, etc.) and others like https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor are doing good work but limited by missing EF features like @alienwareone mentioned (https://github.com/dotnet/SqlClient/issues/465) and generally built on a hacky framework -- query text parsing is the only way they've been able to make this work -- because there are not enough low level EF hooks to make this possible by a third party in a reliable way.

I'm not going to base my business on that sort of foundation, and I need second level caching, so I won't be able to use EF until and unless it's built-in and supported. Others make different calculations, obviously -- there are many people using EF and happy with what it provides. But there are some level of folks that have a different calculus.

**Comment by roji** [comment]
> [...] there are some subset of users such as myself and many others like me who just will not use EF if it doesn't have this key feature built-in and supported.

Unless I'm mistaken, this amounts to saying "I won't use it unless it comes built-in from Microsoft". We sometimes get this: a perfectly valid library exists out there, but some users refuse to use it and ask us to provide a Microsoft solution that duplicates the functionality, only because they don't trust non-Microsoft software. This is IMHO quite a problematic characteristic of the .NET ecosystem, and hinders its progress in general.

At the end of the day, it's simply impractical to ask the EF team to build in everything that any user could want; we're a small team and generally prefer to concentrate on functionality that **must** happen in-the-box, or on features where we have a strong, clear user signal. At this point, I think neither seems true here.

> [...] query text parsing is the only way they've been able to make this work -- because there are not enough low level EF hooks to make this possible by a third party in a reliable way.

This is something you've written before, but I'm still clear on what that concretely means. What exactly is missing in terms of EF hooks? We'd generally rather improve our extensibility to external components to do work, rather than implement everything under the hood.

I'm also not sure why text parsing needs to happen in order to support 2nd-level caching.

> [...] I argue that its people you want as users, people building heavy large LOB apps, people who might not speak up and say "I'm not using EF" but who will quietly go a different way

I understand your argument, but with respect, that's something we frequently hear when someone wants some feature which isn't in EF; it's always possible to make this claim, and it's not really possible to verify whether it's true or not. The fact is that we do get a lot of user signal for other features (see our issues sorted by upvotes), whereas this issue has 9 votes in total.

**Comment by programcsharp** [comment]
> Unless I'm mistaken, this amounts to saying "I won't use it unless it comes built-in from Microsoft". We sometimes get this: a perfectly valid library exists out there, but some users refuse to use it and ask us to provide a Microsoft solution that duplicates the functionality, only because they don't trust non-Microsoft software. This is IMHO quite a problematic characteristic of the .NET ecosystem, and hinders its progress in general.

Yes, that's exactly what I'm saying BECAUSE second level caching is such a fundamental feature and it's easy to get invalidation and concurrency wrong. I'm certainly not going to trust when the only sort of currently maintained solution responds with "works on my machine" and locks the issue when concurrency issues are raised: https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/181 etc. https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/177

RE EF hooks, I see issues such as https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/192 caused by query parsing to figure out if a command is CRUD or not. And looking at the architecture I don't see any hooks at the entity level where they really need to be for a second level caching implementation -- only at the raw command level which is really too low level to optimally handle all of the concurrency and invalidation cases.

I use and love many libraries such as Dapper, StackExchange.Redis, etc. In fact we'll probably end up rolling our own simple ORM + second level caching solution based on those exact libraries since EF doesn't support second level caching.

**Comment by roji** [comment]
> RE EF hooks, I see issues such as https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/192 caused by query parsing to figure out if a command is CRUD or not.

EF Core interceptors are already given context to know whether they're being called for a query, for SaveChanges or for something else; parsing SQL shouldn't be necessary in order to do that.

> I use and love many libraries such as Dapper, StackExchange.Redis, etc. In fact we'll probably end up rolling our own simple ORM + second level caching solution based on those exact libraries since EF doesn't support second level caching.

That's quite an odd reasoning; you're proposing to implement and maintain your own entire ORM + caching solution only because EF doesn't provide one out of the box? If you're unhappy with EFCoreSecondLevelCacheInterceptor, why not contribute to it to make it better, or possibly implement 2nd level caching over EF instead?

**Comment by programcsharp** [comment]
Key word: simple. EF is a lot of things, but simple it is not. I can (and have) written simple second level caching layers on top of simple mappers older than Dapper. I'm confident that I can build and understand a reliable second level caching layer on a simple mapper without all of the other features EF brings. I'm not confident that I can understand all of the moving parts in EF such that I can design such a thing, so if one isn't available I'm going to choose a different solution rather than diving into months of work before a payoff. Even in the simple case where it's easy to see the possible queries and simple to debug through, you run into mistakes and unexpected interactions.

**Comment by roji** [comment]
I must say I find this reasoning strange...

* You're saying you're willing to drop using EF and use e.g. Dapper
* That means you're deciding to forgo using LINQ for querying (and will write your own SQL); to forgo change tracking, migrations, optimistic concurrency and all of the other EF features
* You're letting go of all that because of possible flaws in a well-used community package (EFCoreSecondLevelCacheInterceptor), though I'm still not clear on what those flaws **actually** are; the most I see above is that you don't trust it for some reason, possibly because it's open-source and not from Microsoft.
* You're also not willing to look into doing your own EF interceptor-based implementation because of the theoretical complexity of doing that. This honestly shouldn't be that hard, and there really aren't complex "moving parts" here; just implement an interceptor which can return a reader over cached data instead of executing the command. Any other complexity isn't related to EF and will be the same with Dapper or anything else.

Of course, you'll end up doing whatever is right for you; maybe you don't really care about or need the functionality that EF has to offer, so using Dapper instead is fine. But if EF does present any value to you - which I assume it does as we've been discussing this for a while now - then the reasoning above strikes me as quite odd.

**Comment by programcsharp** [comment]
Mmm... is EFCoreSecondLevelCacheInterceptor actually a mature well-used community package recommended for second level caching by Microsoft? In production scenarios?

Have you read the code? On a quick skim through the code and issues, I see things that look wrong -- like parsing SQL strings rather than checking the context, like locking for every single query: https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/blob/ba4f9b75be59b6a0f47986a15d7ee008cd60c9ec/src/EFCoreSecondLevelCacheInterceptor/SecondLevelCacheInterceptor.cs#L44 and I have my doubts. Does that hold up under load on one machine? That lock certainly wouldn't handle any concurrency questions in any type of clustered/web garden/web farm scenario.

Not to mention because the only hooks EF has got are at the raw data level, you're limited there: https://github.com/dotnet/efcore/issues/5858#issuecomment-629458949 -- it really needs interception capabilities at a higher level -- query/entity.

Ultimately it really comes to problem size vs. benefit. Microsoft feels that second level caching isn't a lot of work to implement and isn't wanted by many folks, so it's not worth implementing as a framework feature. I believe that second level caching is a pretty big task -- caching is hard, particularly distributed caching on top of a transactional store -- and I also think the payoff is pretty large too: most standard LOB of a certain size could get a boost in both speed and scalability from it. Ayende has a good blog post on how big the problem is: https://ayende.com/blog/3112/nhibernate-and-the-second-level-cache-tips

Regardless, I've accomplished my goal here -- raise awareness of second level cache as an important issue.

**Comment by roji** [comment]
> is EFCoreSecondLevelCacheInterceptor actually a mature well-used community package recommended for second level caching by Microsoft?

The point is that you should not be waiting for something that is officially recommended or developed by Microsoft. That's neither scalable nor conducive to an open, vibrant community of .NET software.

> Have you read the code?

I'm neither a user of EFCoreSecondLevelCacheInterceptor, nor have I read its code. But I do know that various users are using it successfully, and also that it's very easy to peer into some code and decide that it's bad. For example, I have no idea what that global lock you're referring to encompasses, and whether it would *actually* slow a real-world scenario down.

What's more interesting is an actual evaluation test, outputting a list *concrete*, user-facing problems you encountered. Otherwise it's very hard to address what you're saying.

> Not to mention because the only hooks EF has got are at the raw data level, you're limited there: https://github.com/dotnet/efcore/issues/5858#issuecomment-629458949 -- it really needs interception capabilities at a higher level -- query/entity.

EF has other interception and extensibility capabilities at higher levels (not that I'm sure why those are necessarily needed for a 2nd-level cache). [Here's the list of those added to 7.0](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#new-and-improved-interceptors-and-events) - we've been adding them to each release. I've still not seen a very concrete example of something you need from EF, and which it doesn't yet provide; until that happens, this is just an ungrounded discussion that can't lead anywhere.

**Comment by programcsharp** [comment]
Microsoft's best practices specifically cover this scenario: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/best-practices?view=aspnetcore-7.0#avoid-blocking-calls says `Do not acquire locks in common code paths. `.

**Comment by roji** [comment]
> The code in question has a singleton lock used for every data access.

The code in question runs after the query completed, so does not encompass I/O. I have no idea what `_processor.ProcessExecutedCommands` encompasses and how much it would take - maybe it's negligible time. In any case, deciding that an entire library is unsuitable because of some lock seen somewhere in the code - well, I'm not sure that's a very comprehensive approach. At the very least I'd expect some engagement with the maintainer around this, possibly an attempt to improve it; after all, that's how OSS works.

**Comment by programcsharp** [comment]
Like I mentioned in my first post, I came here to see if second level caching would ever be supported, as I start to gather options for a transition to a .NET core ORM and provide some feedback as to why it would be cool.

Sounds like we've got an answer here.

**Comment by bzbetty** [comment]
> EF Core interceptors are already given context to know whether they're being called for a query, for SaveChanges or for something else; parsing SQL shouldn't be necessary in order to do that.

EFSecondLevelCache parses table names out of queries / commands in order to compute cache dependencies and when they should be cleared.  Also means it works nicely with libraries like EFBulkChanges which use raw db commands for things like bulk merges.

**Comment by douglasg14b** [comment]
Poking my head in here to support this.

Understandably this isn't necessarily a core feature. By the flow of this discussion and attempts to solve (kludge) the problem, EF Core could make the problem more solvable (In a general way), by providing additional extendability or interception ability?

I didn't know the name of this kind of catching, but I've definitely seemed it out and implemented it myself in previous projects. A clean, non fragile, way to get this behavior seems beneficial 🤔

**Comment by bzbetty** [comment]
> What it's lacking is support for specific types (e.g. [Spatial](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/4), [HierarchyId](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/170)). This would require to make internal constructors public for SqlDataReader ([see issue](https://github.com/dotnet/SqlClient/issues/465)).

With efcore8 I think it now supports hierarchyid, at least I don't see the same errors I used to.

**Comment by JW-akanm** [comment]
@roji 

Edit: Did a bit more digging through the EF Core 10 source.
Would it be possible to:
1) modify the query compilation pipeline to capture the SQL version of the query expression and attach it to the Linq version the first time a query is compiled (in the sense the two are cached together), and
2) modify the QueryCompiler.ExecuteCore function to split the compilation/cached query retrieval and execution steps and add something like a BeforeQueryExecuting interceptor that passes the query cache key, dbcontext, linq expression, and sql expression?
This should provide enough info to decide whether a query is cacheable or not, whether there's already cached data available with a simple cache key check, and cut off further execution at the earliest possible point if so, making it more efficient, and not having to rely on raw SQL string searches and comparisons. 

Pre-edit:
Brainstorming some possibly to likely incorrect ideas after spending a small amount of time dealing with this issue.

One thing that could possibly address the issue with text parsing for table dependencies is if the Sql version of the query expression tree were exposed in the DbCommandInterceptor interface. This would provide a more reliable version of the command to introspect for caching logic (or any other purpose really). 
Exposing the expression cache key for re-use in framework-external query-related caches could also be helpful, since that's being created every time anyway, afaict.
A nice to have is if the interceptor implementations could return the corresponding materialized entity tree without having to re-implement a caching data reader, and presumably skip a lot of the follow up reader interaction and text processing. I don't know if you would need to synchronize the DbContext DbSet collections, since cached entities would be untracked most of the time, but hopefully it wouldn't be difficult to deal with either way.
Then it would be up to the interceptor implementers as to how granular their cache (invalidation) logic needs to be, in terms of what query structure and content represents cacheable results, but at least they would be using a far more reliable source for the query introspection, and through a more stable public interface.

**Comment by roji** [comment]
@JW-akanm what is it that you're referring to with "SQL version" exactly? Keep in mind that whatever we do here needs to be cross-database (i.e. not SQL Server-specific).

**Comment by JW-akanm** [comment]
> [@JW-akanm](https://github.com/JW-akanm) what is it that you're referring to with "SQL version" exactly? Keep in mind that whatever we do here needs to be cross-database (i.e. not SQL Server-specific).

Afaict, the Linq expression tree is transformed into a generic (provider-agnostic) relational SQL expression tree during the compilation process, then possibly further transformed to a provider-specific SQL expression tree, represented via Microsoft.EntityFrameworkCore.Query.SqlExpressions.* expression objects. While trying to understand the timing of this during execution, I created a simple descendant QuerySqlGenerator (actually the provider-specific version of that) to replace the provider generator, and I can see the SQL expression tree visitation logic executing via overrides, kicked off during the execution of the query against the DB (shortly before the DbCommandInterceptor.CommandInitialized event). However, those SQL expression trees seem to be purely internal to the EF framework code and are ephemeral in the sense they are constructed once as part of the Linq expression to SQL command string conversion, but not generated, persisted or used after that, and not available to EF client code.

Edit:
The area of the EF code I'm looking at is a combination of:
Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutorExpression
Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor

**Comment by roji** [comment]
I think that in order to make progress here we'd need a clear expression of exactly what's needed here - what interception point, what inputs and outputs it would have, what point in the query pipeline it would run, and what exactly it would do. The conversation above has been generally a little bit too vague/general.

Also, although EF may currently be lacking a single, clean interception point that corresponds to exactly what people want, you can always replace EF services and insert logic/a visitor at any point you want. This is definitely advanced and more low-level, and it has various drawbacks, but can at the very least be a good way to experiment and come up with a more 1st-class proposal. Though I will mention that many/most of these are behind EF's query caching mechanisms, i.e. will not run with every query execution, so likely not good enough for a 2nd-level cache.

Finally, I'll just mention again that this issue has only received 12 votes since it was open; given that there's very little use interest (compared to everything else we have going on), and given that this is a design-heavy issue that isn't likely to be trivial, I don't see us investing much in this any time soon.
